src/cunumeric/utilities/dispatch.h

```diff
@@ -0,0 +1,192 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include "legate.h"
+#include "cunumeric/cunumeric_c.h"
+
+namespace cunumeric {
+
+template <int DIM>
+struct inner_type_dispatch_fn {
+  template <typename TypeCode, typename Functor, typename... Fnargs>
+  constexpr decltype(auto) operator()(TypeCode code, Functor f, Fnargs&&... args)
+  {
+    switch (code) {
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT1: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT1, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT2: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT2, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT3: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT3, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT4: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT4, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT5: {
```

Comment:
I think this dispatch should abide by `LEGION_MAX_DIM` and avoid instantiating types that will never arise for a given `LEGION_MAX_DIM`. I guess you could check `code <= MAX_TYPE_NUMBER` early on so that you can fall back to the core's dispatch more quickly.